File: daily/21. 合并两个有序链表.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode *flag=new ListNode,*p=flag;
        while(l1 && l2) //直到有链表为空为止
        {
            if(l1->val>l2->val){
                p->next=l2;
                l2=l2->next;
            }
            else{
                p->next=l1;
                l1=l1->next;
            }
            p=p->next;
        }
        p->next=(l1==NULL)?l2:l1;   //把剩余的加上去
        return flag->next;
    }
};

File: daily/链表/链表easy/21. 合并两个有序链表.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var mergeTwoLists = function(l1, l2) {
    // 递归写法 时间O(m+n),空间O(m+n)  m+n为两链表长度之和
    // if(!l1) return l2
    // if(!l2)  return l1
    // if(l1.val<l2.val) {
    //     l1.next = mergeTwoLists(l1.next, l2)
    //     return l1
    // }
    // else {
    //     l2.next = mergeTwoLists(l1, l2.next)
    //     return l2
    // }

    // 非递归写法，时间O(m+n),空间O(1)
    let res = new ListNode(), p = res
    while(l1 && l2) {
        if(l1.val<l2.val) {
            p.next = l1
            l1 = l1.next
        } else {
            p.next = l2
            l2 = l2.next
        }
        p = p.next
    }
    p.next = l1==null ? l2 : l1
    return res.next
};